rfcs/text/0002-rfc-environment.md

```diff
@@ -47,73 +49,118 @@ Stage 2: Include new or updated yml field definitions for all of the essential f
 
+### Usage with APM agents
 
-### APM Agent
+To define `service.environment` on APM agent configurations, we use `-Delastic.apm.environment=staging` or `export ELASTIC_APM_ENVIRONMENT=staging`, see https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-environment
 
-Use `-Delastic.apm.environment=staging` or `export ELASTIC_APM_ENVIRONMENT=staging`, see https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-environment
-
-Note: we may want to adopt the official field name `-Delastic.apm.service_environment=staging`  and `export ELASTIC_APM_SERVICE_ENVIRONMENT=staging`.
+Example
 
 ```
 java -javaagent:/path/to/elastic-apm-agent-<version>.jar \
-     -Delastic.apm.service_name=my-application \
-     -Delastic.apm.environment=staging \
-     -Delastic.apm.server_urls=http://localhost:8200 \
+     -Delastic.apm.service_name=www-frontend \
+     -Delastic.apm.environment=production \
+     -Delastic.apm.server_urls=https://apm-server.my-ecommerce.com:8200 \
      -Delastic.apm.secret_token= \
-     -Delastic.apm.application_packages=org.example \
-     -jar my-application.jar
+     -Delastic.apm.application_packages=com.myecommerce \
+     -jar www-frontend.jar
 ```
 
-### Filebeat.yml
 
+Note: we may want to evolve the Java system property (`-Delastic.apm.environment`) and the environment variable (`ELASTIC_APM_ENVIRONMENT`) to better reflect the ECS field name, it could look like `-Delastic.apm.service_environment` and `ELASTIC_APM_SERVICE_ENVIRONMENT`.
+
+### Usage with Heartbeat
+
+Synthetics tests are the second major use case to qualify the environment of a service.
+An important difference with APM agents is that an Heartbeat agent is likely to simultenaously run synthetics tests on multiple services spread across multiple environments. The qualification of the `service.name`and `service.environment` are defined per `heartbeat.monitor` rather than globally.
 
-Note: we may want to improve the configuration of Beats to not require to define a processor to specify `service.environment`, similarly to the existing solution to define `fields.*` fields.
+The configuration could look like:
 
 ```yaml
+# EXPERIMENT - NOT SUPPORTED CONFIGURATION SNIPPET
+heartbeat.monitors:
+- type: http
+  id: www-frontend-monitor
+  name: Website Frontend Monitor
+  service:
+    name: www-frontend
+    environment: production
+  urls: ["https://www.my-ecommerce-company.com/status"]
+  schedule: '@every 10s'
+  check.response.status: 200
+  timeout: 2s
+- type: http
+  id: www-frontend-monitor-staging
+  name: Website Frontend Monitor - Staging
+  service:
+    name: www-frontend
+    environment: staging
+  urls: ["https://www.staging.my-ecommerce-company.com/status"]
+  schedule: '@every 10s'
+  check.response.status: 200
+  timeout: 5s
+``` 
+
+Note the support in Heartbeat of `service.name` is waiting for https://github.com/elastic/beats/pull/20330
+
+### Usage with Filebeat and Metricbeat
```

Comment:
This section spends a lot of time on what's effectively an example of how to use the "fields" Beats config option. I don't think anyone is actually pushing for this to be adopted, so I suggest we don't mention it at all in this section, but we flesh out this point in the "concerns" section.

So I think we could rewrite this section with a flow somewhat like this:

* contrary to APM and Hearbeat, Filebeat and Metricbeat have very little semantic knowledge out of the box about services, so they rarely populate these fields.
* However in cases where Fb/Mb are monitoring services that are related to services also monitored by APM/Heartbeat, users may benefit from correlating them.
* The way to do so is to populate `service.*` fields explicitly in the relevant Fb/Mb events.

WDYT? Does this follow the intent of the addition of this field?

Reply:
I have updated the RFC accordingly.